pkg/hostpath/hostpath.go

```diff
@@ -166,28 +167,38 @@ func (hp *hostPath) getVolumePath(volID string, volCtx map[string]string) string
 	return filepath.Join(hp.root, volID, podNamespace, podName, podUID, podSA)
 }
 
-func createFile(path string, buf []byte) {
-	file, err := os.Create(path)
-	if err != nil {
-		klog.Errorf("error creating file %s: %s", path, err.Error())
-		return
-	}
-	defer file.Close()
-	file.Write(buf)
-}
-
 func commonUpsertRanger(podPath, filter string, key, value interface{}) bool {
 	if key != filter {
 		return true
 	}
-	buf, err := json.MarshalIndent(value, "", "    ")
-	if err != nil {
-		klog.Errorf("error marshalling: %s", err.Error())
+	payload, _ := value.(Payload)
+	podFileDir := filepath.Join(podPath, fmt.Sprintf("%s", key))
+	if err := os.MkdirAll(podFileDir, os.ModePerm); err != nil {
+		klog.Errorf("unable to make dir %s: %s", podFileDir, err.Error())
 		return true
 	}
-	podFilePath := filepath.Join(podPath, fmt.Sprintf("%s", key))
-	klog.V(4).Infof("create/update file %s", podFilePath)
-	createFile(podFilePath, buf)
+	if payload.ByteData != nil {
+		for dataKey, dataValue := range payload.ByteData {
+			podFilePath := filepath.Join(podFileDir, dataKey)
+			klog.V(4).Infof("create/update file %s", podFilePath)
+			if err := ioutil.WriteFile(podFilePath, dataValue, 0644); err != nil {
+				klog.Errorf("unable to write file %s: %s", podFilePath, err.Error())
```

Comment:
err.Error() here probably includes the file path that it could not write to, what about just returning an error from this (along with the similar ones below) and handling that error where the function is called?  Seems like that might be cleaner and less repetitive code of calling klog to print the error and then returning false?